### DISCUSSION TITLE
[REQUEST] EZSP v9, EZSP v10, and EZSP v11 (EmberZNet Serial Protocol versions 9, 10, and 11) for Silicon Labs EFR32/Ember Zigbee NCP radios

### DISCUSSION THREAD

**Discussion by Hedda** [arendst/Tasmota#18582] [General]
@s-hadinger suggest update Zigbee2Tasmota (Z2T) to support the EZSP serial protocol revisions, e.i. EZSP v9 EZSP v10, and EZSP v11.

https://tasmota.github.io/docs/Zigbee/

Requesting add support for Silicon Labs EZSP v9 (EmberZNet Serial Protocol version 9), EZSP v10 (EmberZNet Serial Protocol version 10), and EZSP v10 (EmberZNet Serial Protocol version 10) to make it compatible with the latest Silabs Zigbee Coordinator firmware releases.

https://github.com/SiliconLabs/gecko_sdk/releases

Silicon Labs Gecko SDK 4.1.x (GSDK 4.1) has been released with it Zigbee EmberZNet SDK 7.1.x.x which has been bumped to EZSP v9

Silicon Labs Gecko SDK 4.2.x (GSDK 4.2) has been released with it Zigbee EmberZNet SDK 7.2.x.x which has been bumped to EZSP v10

Also, Zigbee EmberZNet 7.2.2.2 was been updated to EZSP v11, though understand EZSP v11 is a minor change compared to v10:

https://www.silabs.com/documents/public/release-notes/emberznet-release-notes-7.2.3.0.pdf

https://www.silabs.com/documents/public/release-notes/emberznet-release-notes-7.2.0.0.pdf

https://www.silabs.com/documents/public/release-notes/emberznet-release-notes-7.1.0.0.pdf

Important is that if a user flashes Silabs EmberZNet 7.1.0 or later firmware then that requires at least EZSP v9 support in the application and it breaks backward compatibility with EZSP v8 (which was still supported in Silabs EmberZNet 7.0.x and earlier firmware).

Updating to ZSP adapter implementation is updated to support additions to EZSP v9. See EZSP Reverence guide Rev 4.4 and 4.3

Developers can see EZSP Reference Guide documentation which has most (but not all) of the changes from 6.10.x highlighted here:

https://www.silabs.com/documents/public/user-guides/ug100-ezsp-reference-guide.pdf

 UG100 still mention version 9 as the documentation has not yet been fully updated by Silicon Labs engineers.

Gecko SDK v4.0 (Zigbee EmberZNet SDK 7.0.0.x and previous versions):
https://github.com/SiliconLabs/gecko_sdk/blob/gsdk_4.0/protocol/zigbee/app/util/ezsp/ezsp-protocol.h
* EZSP_PROTOCOL_VERSION = 8

Gecko SDK v4.1 (Zigbee EmberZNet SDK 7.1.0.x):
https://github.com/SiliconLabs/gecko_sdk/blob/gsdk_4.1/protocol/zigbee/app/util/ezsp/ezsp-protocol.h
* EZSP_PROTOCOL_VERSION = 9

Gecko SDK v4.2 (Zigbee EmberZNet SDK 7.2.0.x):
https://github.com/SiliconLabs/gecko_sdk/blob/gsdk_4.2/protocol/zigbee/app/util/ezsp/ezsp-protocol.h
* EZSP_PROTOCOL_VERSION = 10

Gecko SDK v4.2.2 / Gecko SDK v4.2.3 (Zigbee EmberZNet SDK 7.2.2.0 and Zigbee EmberZNet SDK 7.2.3.0):
https://github.com/SiliconLabs/gecko_sdk/blob/gsdk_4.2/protocol/zigbee/app/util/ezsp/ezsp-protocol.h#L33
* EZSP_PROTOCOL_VERSION = 11
** 0x0B hexadecimal = 11 in decimal, added extra parameter added to `pollHandler` in EZSP v11 ("_Updated the ezspPollHandler function with updated input arguments, that required updating the EZSP_PROTOCOL_VERSION to 0x0B._")

PS: The community repositories are offering unofficial firmware image builds for various Silicon Labs based Zigbee adapters:

* https://github.com/xsp1989/zigbeeFirmware/
* https://github.com/skgsergio/silabs-multiprotocol-firmware-zbgw04-usb/tree/main/firmware/ncp
* https://github.com/grobasoz/zigbee-firmware/
* https://github.com/NabuCasa/silabs-firmware
* https://github.com/itead/Sonoff_Zigbee_Dongle_Firmware

PPS: You can now also compile own EmberZNet Zigbee NCP (EZSP) firmware with silabs-firmware-builder originally from Nabu Casa:

* https://github.com/ksjh/silabs-firmware-builder
* https://github.com/darkxst/silabs-firmware-builder
* https://github.com/giannello/silabs-firmware-builder
* https://github.com/NabuCasa/silabs-firmware-builder

**Comment by Hedda** [comment]
FYI, many other open-source project in the DIY home automation community has recently been updated to support EZSP v9, v10, and v11.

The "bellows" radio library for zigpy to interface Silicon Labs EmberZNet EZSP has been updated to support support EZSP v9, v10, and v11:

* [Zigbee Home Automation (ZHA) integration for Home Assistant](https://www.home-assistant.io/integrations/zha) (which as you might know depends on [zigpy](https://github.com/zigpy/zigpy)).
* [Domoticz](https://domoticz.com) via the [Zigbee for Domoticz Plugin](https://www.domoticz.com/wiki/ZigbeeForDomoticz) (see [Domoticz-Zigbee](https://github.com/zigbeefordomoticz/Domoticz-Zigbee), which is also depends on [zigpy](https://github.com/zigpy/zigpy)).
* [Jeedom](https://www.jeedom.com/en/) via [official Zigbee plugin for Jeedom ](https://doc.jeedom.com/en_US/plugins/automation%20protocol/zigbee/) (which is also depends [zigpy](https://github.com/zigpy/zigpy) as well).

The "zigbee-herdsman" library has also already been updated to support EZSP v9, v10, and v11:

* [Zigbee2MQTT](https://www.zigbee2mqtt.io/ ) (that is depending on [zigbee-herdsman](https://github.com/Koenkk/zigbee-herdsman/issues/319)).
* [ioBroker](https://github.com/ioBroker/ioBroker.zigbee) (ioBroker adapter for Zigbee devices that also depends on [zigbee-herdsman](https://github.com/Koenkk/zigbee-herdsman/issues/319)).

The one large project that was lagging behind those others is OpenHAB but its dependencies were at least updated to EZSP v11 last month:

* [OpenHAB](https://www.openhab.org/) via official [OpenHAB ZigBee Binding](https://www.openhab.org/addons/bindings/zigbee/) (that depends on the "[com.zsmartsystems.zigbee](https://github.com/zsmartsystems/com.zsmartsystems.zigbee)" framework library)
  * Note there are also a few other lesser-known projects that also depend on the same "[com.zsmartsystems.zigbee](https://github.com/zsmartsystems/com.zsmartsystems.zigbee)" library.

**Comment by s-hadinger** [comment]
Thanks. Not the highest priority for now unless it brings a major new feature. 

Anyways I can't do anything unless @xsp1989 provides a signed firmware.

**Reply by xsp1989** [reply]
I can provide signed firmware, but you'd better tell me which version you need.

**Reply by Hedda** [reply]
> **Not the highest priority for now** unless it brings a major new feature.

That is understandable. Almost all major new features are related to Multi-PAN support to allow a single radio to run a Zigbee stack (NCP or RCP) and Thread/OpenThread stack or Bluetooth stack concurrently at the same time. 

There are no new major features that are only Zigbee NCP specific, though the stack has been inofficially bumped to Zigbee  R23.

So not really useful unless natively for the  Zigbee2Tasmota (Z2T) project unlesss Tasmota also gets a Thread2Tasmota feature ;)  ...but newer firmware is however useful for users using serial server mode to pass-through the EZSP interface to third-party Zigbee gateway applications such as the ZHA integration in Home Assistant or Zigbee2MQTT.

https://csa-iot.org/newsroom/looking-ahead-industry-trends-and-new-zigbee-features/

Zigbee 3.0 Revision 23 (R23) is next version of the Zigbee PRO standard, expands on existing security by adding new security features: 

* Dynamic Link Key (DLK) joining: Using elliptic curve Diffie-Hellman key negotiation to establish keys and authenticate devices before joining the network 
* Device interview: Using DLK, information is requested to identify the device before accepting it into the network 
* Frame counter synchronization to provide advanced message replay protection
* Network parameter security with PANID and channel management

Newer devices will be able to take advantage of new Zigbee PRO R23 security features if and when network make use of them.
